File: src/main/buttons/iconbuttons/ToggleButton.java
package main.buttons.iconbuttons;

import main.Main;
import processing.core.PImage;

public class ToggleButton extends IconButton {

    private boolean toggled = true;
    private final PImage onImage;
    private final PImage onImageOverlay;
    private final PImage offImage;
    private final PImage offImageOverlay;

    public ToggleButton(String path) {
        super(path);

        //TODO: this is a rough implementation
        this.onImage = Main.INSTANCE.loadImage(path);
        this.onImage.resize(this.getWidth(), this.getHeight());
        this.onImageOverlay = brightenImage(this.onImage);

        this.offImage = Main.INSTANCE.loadImage(path.replace(".png", "_off.png"));
        this.offImage.resize(this.getWidth(), this.getHeight());
        this.offImageOverlay = brightenImage(this.offImage);
    }

    @Override
    protected void drawButton() {
        Main.INSTANCE.image(this.toggled ? this.onImage : this.offImage, this.getX(), this.getY());
    }

    @Override
    protected void getTintedImage() {
        this.tintedImage = this.toggled ? this.onImageOverlay : this.offImageOverlay;
    }

    @Override
    public void clicked() {
        this.toggled = !this.toggled;
    }

    public void setToggled(boolean toggled) {
        this.toggled = toggled;
    }

    public boolean isToggled() {
        return this.toggled;
    }
}

File: src/main/buttons/iconbuttons/ToggleButton.kt
package main.buttons.iconbuttons

import main.Main

open class ToggleButton(path: String) : IconButton(path) {

    var isToggled = true
    //TODO: this is a rough implementation
    private val onImage = Main.loadImage(path)
    private val onImageOverlay = brightenImage(onImage)
    private val offImage = Main.loadImage(path.replace(".png", "_off.png"))
    private val offImageOverlay = brightenImage(offImage)

    override val tintedImage = if (isToggled) onImageOverlay else offImageOverlay

    init {
        onImage.resize(width, height)
        onImageOverlay.resize(width, height)
        offImage.resize(width, height)
        offImageOverlay.resize(width, height)
    }

    override fun drawButton() {
        Main.image(if (isToggled) onImage else offImage, x, y)
    }

    override fun clicked() {
        isToggled = !isToggled
    }
}